page-server: don't set TCP_CORK for tls sockets

rst0git opened the pull request:
This pull request aims to address the errors seen in https://github.com/checkpoint-restore/criu/issues/1380. It looks like this error occurs sometimes after a test the restore phase of a test, when zdtm attempts to kill the page-server and lazy-pages processes.

In particular, when TCP_CORK is set on the socket it tells TCP to wait to remove the cork before sending any packets. As a results, we block the underlying channel, but still send data to gnutls which then produces multiple encryption packets.

This pull request avoids setting TCP_CORK when TLS is used, and fixes a few other minor issues.

adrianreber commented:
Hard to tell if it actually fixes #1380 because it does not happen always, but I am :+1:  merging this to see if #1380 is gone.

avagin commented on criu/page-xfer.c:
```diff
@@ -1389,6 +1389,11 @@ int cr_page_server(bool daemon_mode, bool lazy_dump, int cfd)
+	/*
+	 * We might need a lot of pipes to fetch huge number of pages.
+	 */
+	rlimit_unlimit_nofile();
```
Does the page server splice memory into pipes?

avagin commented on criu/tls.c:
```diff
@@ -70,6 +70,9 @@ ssize_t tls_send(const void *buf, size_t len, int flags)
 			errno = ENOMSG;
 			break;
+		case GNUTLS_E_PREMATURE_TERMINATION:
+			errno = ECONNRESET;
```
where do we terminate the tcp connection?

avagin commented on test/zdtm.py:
```diff
@@ -1886,8 +1886,8 @@ def do_run_test(tname, tdesc, flavs, opts):
                 check_visible_state(t, s, opts)
                 if opts['join_ns']:
                     check_joinns_state(t)
-                t.stop()
```
This is here intentionally. Processes have to be fully functional after restore.

avagin commented:
> In particular, when TCP_CORK is set on the socket it tells TCP to wait to remove the cork before sending any packets. As a result, we block the underlying channel, but still send data to gnutls which then produces multiple encryption packets.

I don't understand this. Could you elaborate when we send data to a tcp sock and when we send data to gnutls?

avagin commented:
TCP_CORK is a kernel feature. For tls, it should look like one packet. From the commit message, I don't understand why it is a problem. Are you able to reproduce the bug? Are you sure that this patch series fix it?

rst0git commented:
> TCP_CORK is a kernel feature.

> Could you elaborate when we send data to a tcp sock and when we send data to gnutls?

TLS encryption/decryption happens in userspace. Instead of directly sending data to the socket with `send()`/`recv()`, we use `gnutls_record_send()` and `gnutls_record_recv()`.

There is `gnutls_record_cork()` and `gnutls_record_uncork()` that could be used as a replacement, but this is unrelated to the failures we have seen in CI.

> From the commit message, I don't understand why it is a problem.

TCP_CORK is not the problem.

> Are you able to reproduce the bug? Are you sure that this patch series fix it?

I'm able to reproduce the error locally and as described in https://github.com/checkpoint-restore/criu/issues/1380. I have updated the GitHub issue with all logs files. I have also been using the CI in my fork for testing.

I moved the fix in a separate pull request with more detailed description: https://github.com/checkpoint-restore/criu/pull/1693
